File: week1/김고은/2108_통계학/2108.js
let fs = require("fs");
let filePath = process.platform === "linux" ? "/dev/stdin" : __dirname + "/input.txt";
let input = fs.readFileSync(filePath).toString().trim().split("\n").map(Number);

let N = input[0];
let numbers = input.slice(1);

let sorted = numbers.sort((a, b) => a - b);

class Calculator {
    constructor (arr) {
        this.arr = arr
    }

    length () {
        return this.arr.length;
    }

    avg () {
        let sum_arr = this.arr.reduce((acc, val) => acc + val, 0);
        let avg = Math.round(sum_arr / this.length());
        return  avg === -0? 0: avg;
    }

    median () {
        return this.arr[Math.floor(this.length() / 2)];
    }

    range () {
        return this.arr[this.length() -1 ] -this.arr[0]; 
    }

    mode() {
        let freq = new Map();

        for (let i of this.arr) {
            freq.set(i, (freq.get(i) || 0) + 1);
        }

        let max_freq = Math.max(...freq.values());

        let modes = [];
        for (let [num, count] of freq.entries()) {
            if (count === max_freq) {
                modes.push(num);
            }
        }

        modes.sort((a, b) => a - b);

        return modes.length > 1 ? modes[1] : modes[0];
    }
};

let rate = new Calculator(sorted);

console.log(rate.avg());
console.log(rate.median());
console.log(rate.mode())
console.log(rate.range());

File: week1/김수빈/2108.py
import sys
from collections import Counter
input = sys.stdin.readline

n = int(input())
numbers = [int(input()) for _ in range(n)]
numbers.sort()

# 1. 산술평균 (반올림)
avg = round(sum(numbers) / n)

# 2. 중앙값
median = numbers[n // 2]

# 3. 최빈값
counter = Counter(numbers)
max_count = counter.most_common(1)[0][1]
candidates = [num for num, count in counter.items() if count == max_count]
candidates.sort()
# 두 번째로 작은 값 사용
mode = candidates[0] if len(candidates) == 1 else candidates[1]

# 4. 범위
diff = numbers[-1] - numbers[0]

sys.stdout.write(str(avg) + "\n")
sys.stdout.write(str(median) + "\n")
sys.stdout.write(str(mode) + "\n")
sys.stdout.write(str(diff) + "\n")
